### ISSUE TITLE
Pickle not working with SolutionArray

### ISSUE THREAD

**Issue by pgkirsch** [convexengineering/gpkit#927]
``` python
In [1]: %run simpleflight.py

In [2]: import pickle

In [3]: with open('sol1.pickle', 'wb') as handle:
   ...:     pickle.dump(sol, handle)
   ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-1c3d7ead98c1> in <module>()
      1 with open('sol1.pickle', 'wb') as handle:
----> 2     pickle.dump(sol, handle)
      3

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in dump(obj, file, protocol)
   1374
   1375 def dump(obj, file, protocol=None):
-> 1376     Pickler(file, protocol).dump(obj)
   1377
   1378 def dumps(obj, protocol=None):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in dump(self, obj)
    222         if self.proto >= 2:
    223             self.write(PROTO + chr(self.proto))
--> 224         self.save(obj)
    225         self.write(STOP)
    226

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    399         else:
    400             save(func)
--> 401             save(args)
    402             write(REDUCE)
    403

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_tuple(self, obj)
    566         write(MARK)
    567         for element in obj:
--> 568             save(element)
    569
    570         if id(obj) in memo:

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    399         else:
    400             save(func)
--> 401             save(args)
    402             write(REDUCE)
    403

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_tuple(self, obj)
    566         write(MARK)
    567         for element in obj:
--> 568             save(element)
    569
    570         if id(obj) in memo:

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in sbatch_setitems(self, items)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _ave(self, obj)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s _reduce_ex(self, proto)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s
 obj)
    399         else:
    400             save(func)
--> 401             save(args)
    402             write(REDUCE)
    403

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s
    566         write(MARK)
    567         for element in obj:
--> 568             save(element)
    569
    570         if id(obj) in memo:

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _
    666         if not self.bin:
    667             for k, v in items:
--> 668                 save(k)
    669                 save(v)
    670                 write(SETITEM)

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in s
    304             reduce = getattr(obj, "__reduce_ex__", None)
    305             if reduce:
--> 306                 rv = reduce(self.proto)
    307             else:
    308                 reduce = getattr(obj, "__reduce__", None)

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\copy_reg.pyc in
     82             dict = None
     83     else:
---> 84         dict = getstate()
     85     if dict:
     86         return _reconstructor, args, dict

TypeError: 'NoneType' object is not callable
...


```

**Comment by pgkirsch** [comment]
@bqpd are you aware of any python wizardry that would allow me to reduce a solution array to a dict?

**Comment by pgkirsch** [comment]
Let me rephrase: any reason that doing the following would be bad?:

```python
In [6]: newdict = {}

In [7]: for k, v in sol.iteritems():
   ...:     newdict[k] = v
   ...:
```

**Comment by pgkirsch** [comment]
Blah just kidding that doesn't help me with my issue

**Comment by bqpd** [comment]
It's pint.

**Comment by bqpd** [comment]
Quantitys don't pickle

**Comment by pgkirsch** [comment]
Ah, right, as you mentioned in #867.

Might we also need to [add a `__getstate__` and `__setstate__` to the VarKey class](http://stackoverflow.com/questions/2049849/why-cant-i-pickle-this-object)?

**Comment by pgkirsch** [comment]
If it were just pint, I think the following would work:

```python
In [4]: newdict = sol # A solution array

In [5]: for k, v in newdict['freevariables'].iteritems():
   ...:         newdict['freevariables'][k] = v.magnitude
   ...:

In [6]: del newdict['cost']

In [7]: del newdict['sensitivities']

In [8]: del newdict['variables']

In [9]: del newdict['constants']

In [10]: import pickle

In [11]: pickle.dump(newdict, open("save.p", "wb"))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-11-f1e00fb347ab> in <module>()
----> 1 pickle.dump(newdict, open("save.p", "wb"))

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in dump(obj, file, protocol)
   1374
   1375 def dump(obj, file, protocol=None):
-> 1376     Pickler(file, protocol).dump(obj)
   1377
   1378 def dumps(obj, protocol=None):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in dump(self, obj)
    222         if self.proto >= 2:
    223             self.write(PROTO + chr(self.proto))
--> 224         self.save(obj)
    225         self.write(STOP)
    226

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    399         else:
    400             save(func)
--> 401             save(args)
    402             write(REDUCE)
    403

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_tuple(self, obj)
    566         write(MARK)
    567         for element in obj:
--> 568             save(element)
    569
    570         if id(obj) in memo:

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    399         else:
    400             save(func)
--> 401             save(args)
    402             write(REDUCE)
    403

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_tuple(self, obj)
    566         write(MARK)
    567         for element in obj:
--> 568             save(element)
    569
    570         if id(obj) in memo:

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    666         if not self.bin:
    667             for k, v in items:
--> 668                 save(k)
    669                 save(v)
    670                 write(SETITEM)

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    304             reduce = getattr(obj, "__reduce_ex__", None)
    305             if reduce:
--> 306                 rv = reduce(self.proto)
    307             else:
    308                 reduce = getattr(obj, "__reduce__", None)

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\copy_reg.pyc in _reduce_ex(self, proto)
     82             dict = None
     83     else:
---> 84         dict = getstate()
     85     if dict:
     86         return _reconstructor, args, dict

TypeError: 'NoneType' object is not callable

```

**Comment by pgkirsch** [comment]
Adding the following two lines to varkey,py

```
$ git diff gpkit/varkey.py
diff --git a/gpkit/varkey.py b/gpkit/varkey.py
index ecd2e93..7fb2108 100644
--- a/gpkit/varkey.py
+++ b/gpkit/varkey.py
@@ -103,6 +103,10 @@ class VarKey(object):
     def __getattr__(self, attr):
         return self.descr.get(attr, None)

+    def __getstate__(self): return self.__dict__
+
+    def __setstate__(self, d): self.__dict__.update(d)
+
     def unitstr(self):
         "Returns latex unitstr"
         us = unitstr(self.units, r"~\mathrm{%s}", "L~")
```

changes the traceback to:
```python
In [3]: newdict = sol

In [4]: for k, v in newdict['freevariables'].iteritems():
   ...:     newdict['freevariables'][k] = v.magnitude
   ...:

In [5]: del newdict['cost']

In [6]: del newdict['sensitivities']

In [7]: del newdict['variables']

In [8]: del newdict['constants']

In [9]: import pickle

In [10]: pickle.dump(newdict, open("save.p", "wb"))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-f1e00fb347ab> in <module>()
----> 1 pickle.dump(newdict, open("save.p", "wb"))

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in dump(obj, file, protocol)
   1374
   1375 def dump(obj, file, protocol=None):
-> 1376     Pickler(file, protocol).dump(obj)
   1377
   1378 def dumps(obj, protocol=None):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in dump(self, obj)
    222         if self.proto >= 2:
    223             self.write(PROTO + chr(self.proto))
--> 224         self.save(obj)
    225         self.write(STOP)
    226

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    423
    424         if state is not None:
--> 425             save(state)
    426             write(BUILD)
    427

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    423
    424         if state is not None:
--> 425             save(state)
    426             write(BUILD)
    427

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    329
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332
    333     def persistent_id(self, obj):

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems,
 obj)
    423
    424         if state is not None:
--> 425             save(state)
    426             write(BUILD)
    427

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save_dict(self, obj)
    653
    654         self.memoize(obj)
--> 655         self._batch_setitems(obj.iteritems())
    656
    657     dispatch[DictionaryType] = save_dict

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in _batch_setitems(self, items)
    667             for k, v in items:
    668                 save(k)
--> 669                 save(v)
    670                 write(SETITEM)
    671             return

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\pickle.pyc in save(self, obj)
    304             reduce = getattr(obj, "__reduce_ex__", None)
    305             if reduce:
--> 306                 rv = reduce(self.proto)
    307             else:
    308                 reduce = getattr(obj, "__reduce__", None)

C:\Users\philippe.kirschen\AppData\Local\Continuum\Anaconda2\lib\copy_reg.pyc in _reduce_ex(self, proto)
     68     else:
     69         if base is self.__class__:
---> 70             raise TypeError, "can't pickle %s objects" % base.__name__
     71         state = base(self)
     72     args = (self.__class__, base, state)

TypeError: can't pickle file objects
```

**Comment by pgkirsch** [comment]
A passable working solution for the meantime:

```python
In [3]: newdict = {}

In [4]: newdict['freevariables'] = {}

In [5]: for k, v in sol['freevariables'].iteritems():
   ...:     newdict['freevariables'][k] = v.magnitude
   ...:

In [6]: import pickle

In [7]: pickle.dump(newdict, open("save.p", "wb"))

```

*NOTE*: This requires the additional two lines in `varkey.py` to work

**Comment by bqpd** [comment]
Looks great. I had get/setstate just operate on the descr dictionary, but it's the same either way.

@pgkirsch, what's the usecase for the pickled solution? (in particular, does it make sense to use varkey keys if you're not using a keydict to store them? I think keydicts will pickle...)

**Comment by pgkirsch** [comment]
My use case is wanting to commit a pickled solution with each code commit, so that the solution of any future model can be compared against it (using something better than just a .sol diff).

Sorry, can you clarify what you mean: what is the descr dictionary and how does it relate here? Also, how does a keydict help me store the solution?

I'm curious because although my solution above seems to work ok when changing source, dynamically adding the exact same methods to VarKey only seems to work for `getstate` and not for `setstate`.

**Comment by pgkirsch** [comment]
Do you mean using the line:
```python
newdict['freevariables'] = KeyDict()
```
...because that runs into the same pickling issues as before.

**Comment by bqpd** [comment]
Oh, KeyDict might need some pickle-proofing too then.

**Comment by pgkirsch** [comment]
I think it's just VarKey that's the problem, because `__getattr__` is overwritten.

**Comment by bqpd** [comment]
get pickle working with
 - [ ] nomial
 - [ ] solutionarray
 - [ ] inequalities
 - [ ] constraintsets?
 - [ ] models?

**Comment by whoburg** [comment]
This is an example of why I think that PosynomialInequality, SignomialInequality, and perhaps other constraints, ought to inherit from `NomialData`. Then A matrix generation, cs, and p_idxs would move to `NomialData`.

This would make a number of gpfit implementations a lot simpler too!

**Comment by bqpd** [comment]
`PosynomialInequality` and `SignomialInequality` don't implement A, cs, p_idxs...

**Comment by whoburg** [comment]
... but they could! (and we need to think of them has having an A for a bunch of implementations in gpfit).

**Comment by bqpd** [comment]
Got it! That's a good plan.
